flask_session_captcha/__init__.py

```diff
@@ -1,97 +1,159 @@
-        # they are most likely open to replay attacks. This addon is built upon flask-session.
-        session_type = app.config.get('SESSION_TYPE')
-        if session_type is None or session_type == "null":
-            raise RuntimeWarning(
-                "Flask-Sessionstore is not set to use a server persistent storage type. This likely means that captchas are vulnerable to replay attacks.")
-        elif session_type == "sqlalchemy":
-            # I have to do this as of version 0.3.1 of flask-session if using sqlalchemy as the session type in order to create the initial database.
-            # Flask-sessionstore seems to have the same problem.
-            app.session_interface.db.create_all()
-
-    def generate(self):
+    def generate(self,*args, **kwargs) -> str:
         """
-        Generates and returns a numeric captcha image in base64 format. 
-        Saves the correct answer in `session[self.session_key]`
-        Use later as:
+            generate captcha with given flags
+
+            :param:
+                numeric: if True generate captcha with numeric only
+                alphabet: if True generate captcha with alphabet
+                punctuation if True generate captcha with punctuation symbols
+
+            if both alphabet and numeric set to True this function generate captcha that contain both alphabet and numeric
 
-        src = captcha.generate()
-        <img src="{{src}}">
         """
-        answer = self.rand.randrange(self.max)
-        answer = str(answer).zfill(self.digits)
+
+        if not kwargs.get("numeric"):
+            numeric = self.include_numeric
+        if not kwargs.get("alphabet"):
+            alphabet = self.include_alphabet
+        if not kwargs.get("punctuation"):
+            punctuation = self.include_punctuation
+
+        # single mode captcha options
+        answer = []
+        if numeric and not alphabet and not punctuation:
+            answer += self.random_number(length=self.length)
+        if alphabet and not numeric and not punctuation:
+            answer += self.random_alphabet(length=self.length)
+        if punctuation and not numeric and not alphabet:
+            answer += self.random_punctuation(length=self.length)
+
+        if len(answer) == 0:  # mix captcha mode
+            selected = {}
+            if punctuation:
+                selected["punctuation"] = self.random_punctuation
+            if numeric:
+                selected["numeric"] = self.random_number
+            if alphabet:
+                selected["alphabet"] = self.random_alphabet
+
+            total = sum([alphabet, numeric, punctuation])
+            each_round = self.length // total
+            for each in selected:
+                answer += (selected[each](length=each_round))
+
+            answer += (self.random_alphabet(length=(self.length - len(answer))))
+            self.shuffle_list(answer)
```

Comment:
Cleaner way to write this code would be using a keyset and `secrets.choice` https://docs.python.org/3/library/secrets.html#secrets.choice 

i.e. something like:
```python
import string
import secrets

keyset = []
if alphabet:
  keyset += string.ascii_letters
if numeric:
  keyset += string.digits
if punctuation:
  keyset += string.punctuation

if not keyset:
  raise Exception("empty keyset")

answer = "".join([secrets.choice(keyset) for _ in range(6)])
```

Reply:
if app config was something like this:
``` python
app.config['CAPTCHA_LENGTH'] = 6
app.config['CAPTCHA_INCLUDE_ALPHABET'] = True
app.config['CAPTCHA_INCLUDE_NUMERIC'] = True
app.config['CAPTCHA_INCLUDE_PUNCTUATION'] = False
```
This method does not specifically guarantee that for a captcha of length 6, there are exactly 3 numeric members and 3 letter members, it can generate a captcha with 5 letters and 1 number.  or ...

But with the current method, it is exactly guaranteed that half of the length of the captcha is  number and half of the captcha is  letter

Reply:
I see. This is adding new behaviour that wasn't really the intent behind the original issue. Should be fine, but this needs to be covered by tests.